Add hydra_rq_launcher

jan-matthis opened the pull request:
## Motivation

This PR adds a launcher plugin for [Redis Queue (RQ)](https://python-rq.org):

> RQ (Redis Queue) is a simple Python library for queueing jobs and processing them in the background with workers.

I've found RQ to be very stable and quick to set up. Jobs can be organised in different queues to manage priorities and it excels at [job monitoring](https://python-rq.org/docs/monitoring/), e.g. through a [web frontend](https://github.com/Parallels/rq-dashboard). 

This grew out of a personal need in a research setting where I had to run lots of small to medium sized independent jobs across multiple nodes. I imagine that this could be a useful plugin and would be glad to make it an official contribution.


### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes


## Test Plan

The structure of the plugin matches the Joblib plugin as closely as possible. Tests pass locally on macOS. As detailed in the `README`, a Redis server would be needed to run this plugin on CI.

Similar to the Joblib plugin, Windows won't be supported [due to limitations of `fork()`](https://python-rq.org/docs/#limitations).

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py:
```diff
@@ -0,0 +1,67 @@
+    password: Optional[str] = "${env:REDIS_PASSWORD}"
+
+
+@dataclass
+class RedisConfDefaults:
+    host: str = "127.0.0.1"
+    port: int = 6379
+    db: Optional[str] = None
+    password: Optional[str] = None
+
```
unused, delete.

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@dataclass
+class RedisConfEnv:
+    host: str = "${env:REDIS_HOST}"
+    port: int = "${env:REDIS_PORT}"
+    db: Optional[str] = "${env:REDIS_DB}"
+    password: Optional[str] = "${env:REDIS_PASSWORD}"
```
This passes mypy just because your interpolated values are accidentally strings.
get used to using SI or II ([doc](https://omegaconf.readthedocs.io/en/latest/structured_config.html#interpolations).)

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@dataclass
+class RedisConfEnv:
```
Rename to RedisConf or RedisLauncherConf.

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py:
```diff
@@ -0,0 +1,67 @@
+    # redis configuration
+    redis: RedisConfEnv = RedisConfEnv()
+    # wait time in seconds when polling results
+    wait_polling: float = 0.1
```
0.1 sec is too aggressive for a defaut. change to 1.

omry commented on plugins/hydra_rq_launcher/README.md:
````diff
@@ -0,0 +1,67 @@
+
+It is good practice not to store connection credentials in code, since they might be accidentally shared. Instead, the default configuration will read them from environment variables. 
+
+On a Linux system with bash, they can be set as follows:
+```commandline
+export REDIS_HOST="localhost"
+export REDIS_PORT="6379"
+export REDIS_DB="0"
+export REDIS_PASSWORD=""
+export REDIS_URL="redis://:$REDIS_PASSWORD@$REDIS_HOST:$REDIS_PORT/$REDIS_DB"
````
mention that by default the plugin is using the environment variables (and document them).

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,146 @@
+) -> JobReturn:
+    """Calls `run_job`
+    """
+    if work_dir is not None:
+        os.chdir(work_dir)
```
why are ou chdiring here?
run_job is supposed to handle it for the job.

jan-matthis replied:
I took this out in a follow-up commit. Reason for originally including this is detailed in my comment below: https://github.com/facebookresearch/hydra/pull/683#issuecomment-644942327

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,146 @@
+        log.info(f"Enqueued {job.get_id()}")
+
+    log.info("Finished enqueuing")
+    if rq_cfg["quit_after_enqueue"]:
+        quit()
```
I am tempted to not support this.
the plugin should definitely not call quit in the middle of processing.
if anything it should raise a special exception that indicates to Hydra that this plugin launched asynchronously and that it should finish the run.

jan-matthis replied:
Agreed that quitting directly isn't clean, so I went for implementing a custom exception. I've found it very helpful to have the option of to not wait for results, e.g., in my use cases so far results are stored to disk on remote servers and I don't need the main processes waiting

omry commented on plugins/hydra_rq_launcher/README.md:
````diff
@@ -0,0 +1,67 @@
+```yaml
+defaults:
+  - hydra/launcher: rq
+```
````
You can show the output of `python foo.py --cfg hydra --package hydra.launcher` as documentation for what the plugin config looks like with one example for overriding from the command line.

Also, link to plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py in the repo for additional documentation of the types.

omry commented on plugins/hydra_rq_launcher/README.md:
```diff
@@ -0,0 +1,67 @@
+# Hydra RQ Launcher
```
follow the structure with the other plugins.
1. README is nearly empty and has a link to a website page.
2. For the website page see what other similar plugins looks like.
(header, structure etc).

omry commented on plugins/hydra_rq_launcher/tests/test_rq_launcher.py:
```diff
@@ -0,0 +1,68 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
```
Is this test actually launches to to a redis server?

jan-matthis commented:
Thanks for the feedback!

I've addressed your points with a follow-up commit:
- Documentation is more extensive and moved from the `README` to the website
- Instead of calling `quit()`, a custom exception (`StopAfterEnqueue`) is raised if `stop_after_enqueue` is `True` (it's `False` by default). I have found it very useful to have this option -- e.g., when scheduling many jobs that will run a couple of days jointly and store results to disk. Perhaps you have an idea for a better name?
- For now, testing is done by using the [unit test strategy outlined in RQ docs](https://python-rq.org/docs/testing/#running-jobs-in-unit-tests): On CI, a mock (fake) Redis server is used and the worker is run in the same thread. This is done by setting an environment variable in `conftest.py`. All tests are passing. This is the simple option and possibly good enough for our purposes here.

If you'd rather have a test scenario with a real Redis database and separate worker processes, we can try implementing this as well -- but it will require more work and a couple of changes:
- In `.circleci/config.yml` we'd need to additionally use `redis` docker images on Linux and use `brew` to install `redis-server` on macOS. It would be easiest to do this directly in the beginning, i.e., globally for all tests
- Workers would need to be spawned when the plugin tests are executing. E.g., invoking `subprocess` calls through `conftest.py` or writing a custom fixture

I have tested locally on macOS whether tests pass when using a Redis server and workers in separate processes -- they do, except for a small issue related to relative sweep subdirs (no issue with absolute sweep subdirs):
- `run_job` is changing working directory based on the sweep subdir
- When the sweep subdir is relative, it is relative to where the `rq worker` process was launched from
- There are a few integration tests that specify the sweep sub dir as a relative directory after copying files to a temporary directory. These tests were initially failing when running workers as separate processes. I launched workers prior to running the tests and workers thus weren't finding the temporary directory
- As a workaround, I originally included a `work_dir` parameter set to `hydra.utils.get_original_cwd()` per default. `os.chdir(work_dir)` was called before running `run_job` ensuring that changing into the relative test sweep subdir was successful. This is the answer to your question here: https://github.com/facebookresearch/hydra/pull/683#discussion_r440489393
- In my follow-up commit I took the `work_dir` plugin parameter out again. It just existed to get tests with relative sweep subdirs to work. In practice, users will just need to be aware that relative sweep subdirs are relative to where `rq worker` was launched from which should be intuitive

omry commented:
Thanks, I will review it again in a day or two.

omry commented on website/docs/plugins/rq_launcher.md:
````diff
@@ -0,0 +1,130 @@
+  - hydra/launcher: rq
+```
+
+The default configuration packaged with the plugin is:
````
I am starting to change how I document the configs for plugins.

Let's try it first in this plugin:
1. remove dataclasses and provide a link to the config class in the repo instead.
2. include the output of:

```
python foo.py --cfg hydra --pacakge hydra.launcher
```

in the doc. try it out. I think it's better to use that.

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+    singleton_state: Dict[Any, Any],
+    redis_job_id: str,
+) -> JobReturn:
+    """Calls `run_job`
+    """
```
this comment is not providing useful information in my opinion. remove?

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+    rq_cfg = launcher.rq
+
+    # Redis configuration
+    redis_cfg = launcher.rq["redis"]
```
try:
```
 launcher.rq.redis
```

jan-matthis replied:
I'd prefer this as well, however, I can't figure out how to get this to work. `launcher.rq` is a `dict` rather than `DictConfig` object. Interestingly, `launcher.config.hydra.launcher.params` is a `DictConfig`, but using this rather than `launcher.rq` seems hacky (?). Maybe you have a suggestion?

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+    redis_cfg = launcher.rq["redis"]
+    is_async = True
+    try:
+        is_async = not redis_cfg["mock"]
```
redis_cfg.mock?

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+    try:
+        is_async = not redis_cfg["mock"]
+    except ValidationError:
+        pass
```
what are you hiding here?
swallowing exceptions can lead to upset stomach.

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+        pass
+    if is_async:
+        connection = Redis(
+            host=redis_cfg["host"],
+            port=redis_cfg["port"],
+            db=redis_cfg["db"],
+            password=redis_cfg["password"],
```
```suggestion
            host=redis_cfg.host,
            port=redis_cfg.port,
            db=redis_cfg.db,
            password=redis_cfg.password,
```

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+    jobs = []
+    singleton_state = Singleton.get_state()
+    log.info(
+        "RQ Launcher ({}) is enqueuing {} jobs".format(
+            ",".join([f"{k}={v}" for k, v in rq_cfg.items()]), len(job_overrides),
+        )
```
prefer fstring when possible.

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+        log.info(f"Enqueued {job.get_id()}")
+
+    log.info("Finished enqueuing")
+    if rq_cfg["stop_after_enqueue"]:
+        raise StopAfterEnqueue
```
Can you show me the console output when this happens?
what does it look like to the user?

jan-matthis replied:
Sure, this is how it would look like:
```
[2020-06-19 16:20:52,189][HYDRA] RQ Launcher is enqueuing 3 job(s) in queue : default
[2020-06-19 16:20:52,190][HYDRA] Sweep output dir : multirun/2020-06-19/16-20-52
[2020-06-19 16:20:52,190][HYDRA] 	#1 : task=1
[2020-06-19 16:20:52,216][HYDRA] Enqueued 3ae6c637-529a-43cd-8664-6ce419d2ace4
[2020-06-19 16:20:52,216][HYDRA] 	#2 : task=2
[2020-06-19 16:20:52,240][HYDRA] Enqueued 50408b07-51f4-42ce-b5c7-f586d17f9d61
[2020-06-19 16:20:52,240][HYDRA] 	#3 : task=3
[2020-06-19 16:20:52,263][HYDRA] Enqueued e7118216-aff5-479f-bf1d-9f34f47807d0
[2020-06-19 16:20:52,264][HYDRA] Finished enqueuing
Traceback (most recent call last):
  File "example/my_app.py", line 19, in <module>
    my_app()
  File "/Users/jm/GitHub/jan-matthis/hydra/hydra/main.py", line 37, in decorated_main
    strict=strict,
  File "/Users/jm/GitHub/jan-matthis/hydra/hydra/_internal/utils.py", line 270, in run_hydra
    lambda: hydra.multirun(
  File "/Users/jm/GitHub/jan-matthis/hydra/hydra/_internal/utils.py", line 188, in run_and_report
    return func()
  File "/Users/jm/GitHub/jan-matthis/hydra/hydra/_internal/utils.py", line 273, in <lambda>
    overrides=args.overrides,
  File "/Users/jm/GitHub/jan-matthis/hydra/hydra/_internal/hydra.py", line 127, in multirun
    return sweeper.sweep(arguments=task_overrides)
  File "/Users/jm/GitHub/jan-matthis/hydra/hydra/_internal/core_plugins/basic_sweeper.py", line 113, in sweep
    results = self.launcher.launch(batch, initial_job_idx=initial_job_idx)
  File "/Users/jm/GitHub/jan-matthis/hydra/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/rq_launcher.py", line 43, in launch
    launcher=self, job_overrides=job_overrides, initial_job_idx=initial_job_idx
  File "/Users/jm/GitHub/jan-matthis/hydra/plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py", line 133, in launch
    raise StopAfterEnqueue
hydra_plugins.hydra_rq_launcher._core.StopAfterEnqueue
```

jan-matthis replied:
FR is open here: https://github.com/facebookresearch/hydra/issues/692

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/config.py:
```diff
@@ -0,0 +1,64 @@
+@dataclass
+class RedisConf:
+    # host address
+    host: str = II("env:REDIS_HOST")
+    # port
+    port: int = II("env:REDIS_PORT")
+    # database, None for default
+    db: Optional[str] = II("env:REDIS_DB")
+    # password, None for no password
+    password: Optional[str] = II("env:REDIS_PASSWORD")
+    # if True, runs without redis server in single thread for testing purposes
+    mock: Optional[bool] = II("env:REDIS_MOCK")
```
The env resolver will fail if the environment variables are not present.
you can add a default value if it makes sense:
```python
II("env:REDIS_HOST,localhost")
```

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/_core.py:
```diff
@@ -0,0 +1,166 @@
+
+    sweep_config = config_loader.load_sweep_config(config, list(overrides))
+    with open_dict(sweep_config):
+        sweep_config.hydra.job.id = redis_job_id
```
Good :)

omry commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/rq_launcher.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class RQLauncher(Launcher):
+    def __init__(self, **kwargs: Any) -> None:
```
I think params is a better name here.
```suggestion
    def __init__(self, **params: Any) -> None:
```

omry commented:
I responded to the changes in the code, but let me put some additional inline comments here:

>     * Testing is now done by using the [unit test strategy outlined in RQ docs](https://python-rq.org/docs/testing/#running-jobs-in-unit-tests): On CI, a mock (fake) Redis server is used and the worker is run in the same thread. This is done by setting an environment variable in `conftest.py`. All tests are passing. This is the simple option and possibly good enough for our purposes here.

I think this is a good enough solution.

> If you'd rather have a test scenario with a real Redis database and separate worker processes, we can try implementing this as well -- but it will require more work and a couple of changes:

For now there is no support for per-plugin circleci image so it's going to be pretty hard to do something better.
Once we have that I am sure we will be able to make something clean work.
We can revisit this later.
 
> I have tested locally on macOS whether tests pass when using a Redis server and workers in separate processes -- they do, except for a small issue related to relative sweep subdirs (no issue with absolute sweep subdirs):
> 
>     * `run_job` is changing working directory based on the sweep subdir
> 
>     * When the sweep subdir is relative, it is relative to where the `rq worker` process was launched from
> 
>     * There are a few integration tests that specify the sweep sub dir as a relative directory after copying files to a temporary directory. These tests were initially failing when running workers as separate processes. I launched workers prior to running the tests and workers thus weren't finding the relative subdir inside the temporary directory
> 
>     * As a workaround, I originally included a `work_dir` parameter set to `hydra.utils.get_original_cwd()` (of the main process) per default. `os.chdir(work_dir)` was called on the worker before running `run_job` ensuring that changing into the relative test sweep subdir was successful. This is the answer to your question here: [#683 (comment)](https://github.com/facebookresearch/hydra/pull/683#discussion_r440489393)
> 
>     * In my follow-up commit I took the `work_dir` plugin parameter out again. It just existed to get tests with relative sweep subdirs to work when testing with separate worker processes
> 
>     * Tests now use a worker running in the same thread, so relative sweep subdirs work fine
> 
>     * In practice, users will just need to be aware that relative sweep subdirs are relative to where `rq worker` was launched from -- I guess this should be intuitive

* Is it documented?
* does the basic example override hydra.sweep.dir? you can make it relative to /tmp for example.

You can actually make it an error in the plugin:
If the user is providing a relative job dir you can error out.
This problem is not unique to redis, the same problem would happen with submitit plugin.

jan-matthis commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+from .rq_launcher import RQLauncher
+
+__all__ = ["RQLauncher"]
```
Thanks for catching this! Fixed

jan-matthis commented:
Thanks for the feedback! I have addressed your comments in the latest round of commits. I've replied to two of your inline comments above and was able to resolve the other ones.

I have added a very short news fragment in `NEWS.md`. The convention so far seems to be that headings include version numbers, so I've titled it `1.0.0-rc1 (2020-06-19)` for now. 

The website docs are updated following your suggestion of using ` --cfg hydra --package hydra.launcher`. I did copy-paste the comments from `config.py` into the output (same line) to have more complete documentation on the website. If you prefer, we can leave those out, use the exact script output, and refer to the `config.py` file instead.

Following the rebase on master, LGTM fails saying that the PR has merge conflict. However, just below that I get a green tick mark "This branch has no conflicts with the base branch". Should I do something about it, or this is just a problem LGTM has?

jan-matthis commented:
Ah, and I did not address this suggestion so far:

> You can actually make it an error in the plugin:
> If the user is providing a relative job dir you can error out.
> This problem is not unique to redis, the same problem would happen with submitit plugin.

Since relative sweep dirs are not leading to errors, but can cause to unexpected behaviour, I wanted to suggest to raise a warning rather than throwing an error. E.g., something along these lines:
```python
if not sweep_dir.is_absolute(): 
    log.warn("Using relative sweep dir. Please be aware that the directory will be relative to where workers are started from.")
```

omry commented on NEWS.md:
```diff
@@ -1,3 +1,12 @@
+Plugins
+-------
+
+- Add [Redis Queue (RQ)](https://python-rq.org/) Launcher plugin ([Jan-Matthis](https://github.com/jan-matthis))
```
ah, no.
read [news entries](https://hydra.cc/docs/development/contributing#news-entries) in the developer guide.

jan-matthis replied:
Ah! It's fixed now

omry commented:
> Thanks for the feedback! I have addressed your comments in the latest round of commits. I've replied to two of your inline comments above and was able to resolve the other ones.
> 
> I have added a very short news fragment in `NEWS.md`. The convention so far seems to be that headings include version numbers, so I've titled it `1.0.0-rc1 (2020-06-19)` for now.

Commented on this one, that's not how you do it :)
 
> The website docs are updated following your suggestion of using ` --cfg hydra --package hydra.launcher`. I did copy-paste the comments from `config.py` into the output (same line) to have more complete documentation on the website. If you prefer, we can leave those out, use the exact script output, and refer to the `config.py` file instead.

I'll take a look at the docs again.
I am still trying to figure out how to best document plugins with structured configs.

> Following the rebase on master, LGTM fails saying that the PR has merge conflict. However, just below that I get a green tick mark "This branch has no conflicts with the base branch". Should I do something about it, or this is just a problem LGTM has?

If GitHub is happy, LGTM can GTFO :).

I will review your other changes soon.

omry commented:
looks like you do have a merge conflict after all.

I also seen unrelated changed showing up in your diff from the testing change in master, probably due to a merge but not sure.
see if you can get them out of the diff, possibly by creating a clean branch on top of master and cherry picking your actual changes into it, then force pushing it into this branch in origin. (or into a a new pull request if you are out of options).

omry commented on plugins/hydra_rq_launcher/tests/test_rq_launcher.py:
```diff
@@ -0,0 +1,68 @@
+            {
+                "defaults": [
+                    {"hydra/launcher": "rq"},
+                    {"hydra/hydra_logging": "hydra_debug"},
+                    {"hydra/job_logging": "disabled"},
```
I know this is copy paste.
can you check what happens if you don't override the logging?
does it print new shit when you run the test or is it okay?

jan-matthis replied:
If I take out lines L39-40 tests pass as before, no extra output is printed

jan-matthis replied:
Changed the tests to not override the logging

omry replied:
let's do it.

omry commented on website/docs/plugins/rq_launcher.md:
```diff
@@ -0,0 +1,95 @@
+# @package hydra.launcher
+cls: hydra_plugins.hydra_rq_launcher.rq_launcher.RQLauncher
+params:
+  enqueue:
+    job_timeout: null                  # maximum runtime of the job before it's interrupted and marked as failed (in sec)
+    ttl: null                          # maximum queued time before the job before is discarded (in sec)
+    result_ttl: 864000                 # how long successful jobs and their results are kept (in sec), default: 10 days
+    failure_ttl: 8640000               # specifies how long failed jobs are kept (in sec), default: 100 days
+    at_front: false                    # place job at the front of the queue, instead of the back
+    job_id: null                       # job id, will be overidden automatically by a uuid unless specified explicitly
+    description: null                  # description, will be overidden automatically unless specified explicitly
+  queue: default                       # queue name
+  redis:
+    host: ${env:REDIS_HOST,localhost}  # host address via REDIS_HOST environment variable, default: localhost
+    port: ${env:REDIS_PORT,6379}       # port via REDIS_PORT environment variable, default: 6379
+    db: ${env:REDIS_DB,0}              # database via REDIS_DB environment variable, default: 0
+    password: ${env:REDIS_PASSWORD,}   # password via REDIS_PASSWORD environment variable, default: no password
+    mock: ${env:REDIS_MOCK,False}      # switch to run without redis server in single thread, for testing purposes only
+  stop_after_enqueue: false            # stop after enqueueing by raising custom exception
+  wait_polling: 1.0                    # wait time in seconds when polling results
```
config design question:
Why separate the redis and the enqeue parameters?

jan-matthis replied:
Redis parameters hold connection information for the database. In enqueue parameters are parameters that are passed to `RQ`'s enqueue method (https://python-rq.org/docs/#enqueueing-jobs)

omry replied:
If it is well structured you can combine the two structured for config purposes and just have a bit more code when you are initializing the queue or calling enqueue.

your call.

omry commented on website/docs/plugins/rq_launcher.md:
```diff
@@ -0,0 +1,95 @@
+# @package hydra.launcher
+cls: hydra_plugins.hydra_rq_launcher.rq_launcher.RQLauncher
+params:
+  enqueue:
+    job_timeout: null                  # maximum runtime of the job before it's interrupted and marked as failed (in sec)
+    ttl: null                          # maximum queued time before the job before is discarded (in sec)
+    result_ttl: 864000                 # how long successful jobs and their results are kept (in sec), default: 10 days
+    failure_ttl: 8640000               # specifies how long failed jobs are kept (in sec), default: 100 days
+    at_front: false                    # place job at the front of the queue, instead of the back
+    job_id: null                       # job id, will be overidden automatically by a uuid unless specified explicitly
+    description: null                  # description, will be overidden automatically unless specified explicitly
+  queue: default                       # queue name
```
I know adding comments is making it more useful, but:
1. the comments are in config.py.
2. it makes it harder to update, can't just copy-paste the output of the --cfg command,

For now let's keep this.

omry commented on website/docs/plugins/rq_launcher.md:
```diff
@@ -0,0 +1,95 @@
+
+[HYDRA] RQ Launcher is enqueuing 5 job(s) in queue : default
+[HYDRA] Sweep output dir : multirun/2020-06-15/18-00-00
+[HYDRA] 	#1 : task=1
+[HYDRA] Enqueued 13b3da4e-03f7-4d16-9ca8-cfb3c48afeae
```
Those prints can be cleared.
maybe reverse their order. (even if it means moving the params print to after the job is submitted).

jan-matthis commented on plugins/hydra_rq_launcher/hydra_plugins/hydra_rq_launcher/rq_launcher.py:
```diff
@@ -0,0 +1,44 @@
+        self.config_loader: Optional[ConfigLoader] = None
+        self.task_function: Optional[TaskFunction] = None
+
+        self.rq = params
```
Great, thanks a lot for the thorough explanation!

jan-matthis commented:
Okay! I rebased and squashed to get rid of the merge conflict. All comments should be addressed.

omry commented on website/docs/plugins/rq_launcher.md:
```diff
@@ -0,0 +1,95 @@
+[![Example application](https://img.shields.io/badge/-Example%20application-informational)](https://github.com/facebookresearch/hydra/tree/master/plugins/hydra_rq_launcher/example)
+[![Plugin source](https://img.shields.io/badge/-Plugin%20source-informational)](https://github.com/facebookresearch/hydra/tree/master/plugins/hydra_rq_launcher)
+
+The RQ Launcher plugin provides a launcher for distributed execution and job queuing based on [Redis Queue (RQ)](https://python-rq.org).
```
Can you add a sentence here about about the advantages/disadvantages over the joblib launcher?

omry commented:
for some reason this is again not running against circleci, any idea?

omry commented:
fyi:
I just  landed yamllint, be sure to run `yamllint .` on your  tree to confirm it's passing (it will fail CI otherwise).

omry commented:
rebase first though.

jan-matthis commented:
Everything should be addressed now:
- Included a comparison to the Joblib launcher in [the second paragraph](https://github.com/facebookresearch/hydra/blob/a4c023178eac47a6945784f15ea08df07a26a215/website/docs/plugins/rq_launcher.md) of the website docs
- `flake8` and `black` were in conflict [in this line](https://github.com/facebookresearch/hydra/blob/a4c023178eac47a6945784f15ea08df07a26a215/plugins/hydra_rq_launcher/tests/test_rq_launcher.py#L40), while `black` wants a whitespace after the first comma, flake8 disagrees. Added `noqa: E231`
- A deprecation warning raised by RQ (https://github.com/rq/rq/issues/1244) trigged CI failures (due to `-Werror`), for the time being those [warnings are filtered out](https://github.com/facebookresearch/hydra/blob/a4c023178eac47a6945784f15ea08df07a26a215/plugins/hydra_rq_launcher/tests/test_rq_launcher.py#L23-L25)

All CI tests are passing except for `py36_win` (timeout). This should be unrelated to the plugin. As visible from the `py37_win` and `py38_win` logs, `nox` correctly detects that the plugin does not support Windows (`Deselecting hydra_rq_launcher: Incompatible OS Windows. Supports [MacOS, Linux]`) and skips testing the launcher.

omry commented:
> * `flake8` and `black` were in conflict [in this line](https://github.com/facebookresearch/hydra/blob/a4c023178eac47a6945784f15ea08df07a26a215/plugins/hydra_rq_launcher/tests/test_rq_launcher.py#L40), while `black` wants a whitespace after the first comma, flake8 disagrees. Added `noqa: E231`

Can you remove the trailing comma? I think it will fix it.

> * A deprecation warning raised by RQ (see open issue [rq/rq#1244](https://github.com/rq/rq/issues/1244)) trigged CI failures (due to `-Werror`), for the time being those [warnings are filtered out](https://github.com/facebookresearch/hydra/blob/a4c023178eac47a6945784f15ea08df07a26a215/plugins/hydra_rq_launcher/tests/test_rq_launcher.py#L23-L25) to prevent test failure

Sweet, didn't know you can do that. Thanks!
cc @jieru-hu 


> All CI tests are passing except for `py36_win` (timeout). This should be unrelated to the plugin. As visible from the `py37_win` and `py38_win` logs, `nox` correctly detects that the plugin does not support Windows (`Deselecting hydra_rq_launcher: Incompatible OS Windows. Supports [MacOS, Linux]`) and skips testing the launcher.

I increased the timeout for it for now, can you rebase and try again?

jan-matthis commented:
> Can you remove the trailing comma? I think it will fix it.

That fixed it :)

> I increased the timeout for it for now, can you rebase and try again?

Great, all working now!

omry commented:
Great, thanks!
